R-package/inst/make-r-def.R

```diff
@@ -0,0 +1,98 @@
+# [description]
+#     Create a definition file (.def) from a .dll file, using objdump.
+#
+# [usage]
+#
+#     Rscript make-r-def.R something.dll something.def
+#
+# [references]
+#    * https://www.cs.colorado.edu/~main/cs1300/doc/mingwfaq.html
+
+args <- commandArgs(trailingOnly = TRUE)
+
+IN_DLL_FILE <- args[[1L]]
+OUT_DEF_FILE <- args[[2L]]
+DLL_BASE_NAME <- basename(IN_DLL_FILE)
+
+message(sprintf("Creating '%s' from '%s'", OUT_DEF_FILE, IN_DLL_FILE))
+
+# system() will not raise an R exception if the process called
+# fails. Wrapping it here to get that behavior.
+#
+# system() introduces a lot of overhead, at least on Windows,
+# so trying processx if it is available
+.pipe_shell_command_to_stdout <- function(command, args, out_file) {
+    has_processx <- suppressMessages({
+      suppressWarnings({
+        require("processx")  # nolint
+      })
+    })
+    if (has_processx) {
+        p <- processx::process$new(
+            command = command
+            , args = args
+            , stdout = out_file
+            , windows_verbatim_args = FALSE
+        )
+        invisible(p$wait())
```

Comment:
Does it really return exit status?
https://processx.r-lib.org/#exit-statuses

Please unify all functions for checking exit statuses:
https://github.com/microsoft/LightGBM/blob/ce95d9ca8aab8fd007a2682d4cc80157b4e1b801/build_r.R#L26
https://github.com/microsoft/LightGBM/blob/3caf359a79754825aac17ac77d557d0b7288c70e/R-package/src/install.libs.R#L23

Reply:
this function isn't intended to return the exit code. It runs the command and throws an error if the command fails. `p$wait()` will raise an R error if the command fails.

`processx::proocess$new()` might do this:

![image](https://user-images.githubusercontent.com/7608904/84102295-b0596c00-a9d5-11ea-962d-246a59abdbaa.png)

or if it takes longer to run, `p$wait()` will do it (I couldn't easily reproduce that).

Reply:
could you re-phrase? I don't understand what you're asking me to do with with `Please  unify all functions for checking exist statuses`.

I am still opposed to having one shared function used by all three of these scripts. I think it's important that the scripts are self-contained as much as possible.

Reply:
> It runs the command and throws an error if the command fails.

Got it!

> I am still opposed to having one shared function used by all three of these scripts.

Yeah, I remember! I mean, use the same code for all three functions. Or do you keep the difference between them intentionally?

Reply:
The functions in `install.libs.R` and `build_r.R` are identical because they do the same thing: run a shell command, raise en error on non-0 exit codes if `strict = TRUE`, and return the exit code.

This function has a different purpose because it is running a command and redirecting the `stdout` of that command to a file. I don't want to add `out_file` as an argument to the function in `install.libs.R` and `build_r.R` (where it isn't needed) or add the option `strict` here (where it will always be `TRUE`.

Reply:
OK, I see! Thanks a lot for the explanation!